libraries/splunk_app.rb

```diff
@@ -126,13 +174,80 @@ def create_splunk_directory(path)
         dir.group(node['splunk']['group'])
         dir.mode('0755')
         dir.run_action(:create)
-        new_resource.updated_by_last_action(dir.updated_by_last_action?)
+        new_resource.updated_by_last_action(true) if dir.updated_by_last_action?
+      end
+
+      def download_and_install # rubocop:disable PerceivedComplexity, CyclomaticComplexity
+        return if new_resource.url.nil? || new_resource.url.empty?
+
+        expected_version = CernerSplunk::AppVersion.new new_resource.version
+        installed_version = CernerSplunk::AppVersion.new @current_resource.version
+
+        return if expected_version.type && expected_version == installed_version
+        Chef::Log.warn "Expected version not specified for #{new_resource.app}." unless expected_version.type
+        fail "Expecting to install prerelease on top of same released version for #{new_resource.app}" if expected_version.type == :prerelease && installed_version.type == :base && expected_version.base == installed_version.base
+
+        # In whyrun mode, the rest is not safe.
+        return if whyrun_mode?
+
+        filename = "#{Chef::Config[:file_cache_path]}/#{new_resource.app}.tgz"
+
+        download = Chef::Resource::RemoteFile.new(filename, run_context)
+        download.source(new_resource.url)
+        download.run_action(:create)
+        download.backup(false)
+
+        install_from_tar filename, expected_version, installed_version
+      ensure
+        download.run_action(:delete) if download
+      end
+
+      def install_from_tar(filename, expected_version, installed_version) # rubocop:disable PerceivedComplexity, CyclomaticComplexity
+        tarfile = CernerSplunk::TarBall.new(filename, prefix: new_resource.app, user: node['splunk']['user'], group: node['splunk']['group'])
+
+        fail "Downloaded tarball from '#{new_resource.url}' does not contain an app named '#{new_resource.app}'" if tarfile.num_files == 0
+
+        fail "Downloaded tarball for '#{new_resource.app}' has local entries" unless tarfile.count { |p, _| p.match %r{^[^/]+/local/.+} } == 0
+
+        app_conf = CernerSplunk::Conf.parse_string tarfile.get_file('default/app.conf')
+        tar_version = CernerSplunk::AppVersion.new((app_conf['launcher'] || {})['version'])
+
+        fail "Bad Developer!!!! The downloaded tar for #{new_resource.app} does not contain a version in app.conf!" if tar_version.type.nil?
+
+        if expected_version.type && tar_version != expected_version
+          fail "Expected version #{expected_version} does not match tar version #{tar_version} for #{new_resource.app}" unless expected_version.type == :base && tar_version.base == expected_version.base
+        end
+
+        return if tar_version == installed_version
```

Comment:
Yeah, the additional complexity of looking at the build number is probably not necessary.   I'm ok with the logic you have.  I think we can change our snapshot tarball builds to have a more dynamic pre-release version number component in the app.conf and call that good.